File: 2583-divide-nodes-into-the-maximum-number-of-groups/2583-divide-nodes-into-the-maximum-number-of-groups.py
class Solution(object):
    def magnificentSets(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        graph = [[] for _ in range(n + 1)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        color = [-1]*(n+1)
        
        # Check bipartite and also group nodes by component
        def bfs_check_bipartite(start):
            queue = deque([start])
            color[start] = 0
            comp_nodes = [start]
            while queue:
                u = queue.popleft()
                for w in graph[u]:
                    if color[w] == -1:
                        color[w] = 1 - color[u]
                        comp_nodes.append(w)
                        queue.append(w)
                    elif color[w] == color[u]:
                        return False, []
            return True, comp_nodes

        # We'll store components
        components = []
        for i in range(1, n+1):
            if color[i] == -1:
                is_bip, comp_nodes = bfs_check_bipartite(i)
                if not is_bip:
                    return -1
                components.append(comp_nodes)

        # Function to run BFS restricted to a component's nodes and return max distance
        def max_bfs_distance_from_node(start, comp_set):
            dist = {start:0}
            q = deque([start])
            visited = {start}
            max_dist = 0
            while q:
                u = q.popleft()
                for w in graph[u]:
                    if w in comp_set and w not in visited:
                        visited.add(w)
                        dist[w] = dist[u] + 1
                        max_dist = max(max_dist, dist[w])
                        q.append(w)
            return max_dist

        total_groups = 0
        for comp_nodes in components:
            comp_set = set(comp_nodes)
            # Try BFS from each node in the component to find the max layering
            best_layers = 0
            for node in comp_nodes:
                d = max_bfs_distance_from_node(node, comp_set)
                # layers = distance + 1
                if d + 1 > best_layers:
                    best_layers = d + 1
            total_groups += best_layers

        return total_groups

File: 2583-divide-nodes-into-the-maximum-number-of-groups/2583-divide-nodes-into-the-maximum-number-of-groups.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
function magnificentSets(n, edges) {
    // Build adjacency list
    const graph = Array.from({ length: n + 1 }, () => []);
    for (const [u, v] of edges) {
        graph[u].push(v);
        graph[v].push(u);
    }

    // Find connected components using DFS
    const visited = new Set();
    const components = [];

    function dfs(node, component) {
        if (visited.has(node)) return;
        visited.add(node);
        component.push(node);
        for (const nei of graph[node]) {
            dfs(nei, component);
        }
    }

    // Get all components
    for (let i = 1; i <= n; i++) {
        if (!visited.has(i)) {
            const component = [];
            dfs(i, component);
            components.push(component);
        }
    }

    // BFS to find max groups for a component starting from a specific node
    function bfs(start) {
        const dist = new Array(n + 1).fill(-1);
        const queue = [start];
        dist[start] = 0;
        let maxGroup = 1;

        // Check bipartite property and find max group
        while (queue.length) {
            const node = queue.shift();
            maxGroup = Math.max(maxGroup, dist[node] + 1);

            for (const nei of graph[node]) {
                if (dist[nei] === -1) {
                    dist[nei] = dist[node] + 1;
                    queue.push(nei);
                } else if (Math.abs(dist[nei] - dist[node]) !== 1) {
                    return -1;
                }
            }
        }

        return maxGroup;
    }

    // Process each component
    let totalGroups = 0;

    for (const component of components) {
        let maxComponentGroups = -1;

        // Try each node as starting point
        for (const node of component) {
            const groups = bfs(node);
            if (groups === -1) return -1;
            maxComponentGroups = Math.max(maxComponentGroups, groups);
        }

        totalGroups += maxComponentGroups;
    }

    return totalGroups;
}